File: backend/app/schemas/v2/staff_companion_assignment.py
"""StaffCompanionAssignment v2 schemas (W10-BE1).

同行スタッフ割当 API のリクエスト / レスポンス型定義。

曜日 (weekday: 0=Mon ... 6=Sun) × 午前/午後/終日 (part: am/pm/full) の
単位で新人スタッフ (trainee) への同行スタッフ (companion) を管理する。
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# 同行パート
StaffCompanionPart = Literal["am", "pm", "full"]


class StaffCompanionAssignmentV2Base(BaseModel):
    """同行スタッフ割当 基底スキーマ."""

    model_config = ConfigDict(extra="forbid")

    weekday: int = Field(ge=0, le=6, description="曜日 (0=月 ... 6=日)")
    part: StaffCompanionPart = Field(description="午前 (am) / 午後 (pm) / 終日 (full)")
    companion_staff_id: UUID = Field(description="同行スタッフの staff.id")


class StaffCompanionAssignmentV2Read(StaffCompanionAssignmentV2Base):
    """GET レスポンス用スキーマ."""

    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: UUID
    trainee_staff_id: UUID
    created_at: datetime
    updated_at: datetime


class StaffCompanionAssignmentsBulkPut(BaseModel):
    """PUT /staff/{id}/companion-assignments リクエスト.

    1 TX で当該 trainee_staff_id の全既存削除 → items を INSERT。

    バリデーション (API 層で実施):
      - 同一 (weekday, part) 重複 → 422
      - companion_staff_id が trainee 自身 → 422
      - items は 0..14 件 (7 曜日 × 3 part)
    """

    model_config = ConfigDict(extra="forbid")

    items: list[StaffCompanionAssignmentV2Base] = Field(
        default_factory=list,
        description="同行スタッフ割当リスト (0..14 件)",
    )

File: frontend/lib/schemas/v2/staff_companion_assignment.ts
/**
 * StaffCompanionAssignment v2 zod schemas (W10-FE1).
 *
 * 設計仕様書 §3.5.x / §4.2.x に基づく同行スタッフ割当スキーマ。
 * Backend `backend/app/schemas/v2/staff_companion_assignment.py` と完全一致。
 *
 * 曜日 (weekday: 0=Mon ... 6=Sun) × パート (am/pm/full) の単位で
 * 新人スタッフ (trainee) への同行スタッフ (companion) を管理する。
 */
import { z } from 'zod';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

export const STAFF_COMPANION_PARTS = ['am', 'pm', 'full'] as const;
export type StaffCompanionPart = (typeof STAFF_COMPANION_PARTS)[number];

// ---------------------------------------------------------------------------
// Base / Read schemas
// ---------------------------------------------------------------------------

export const staffCompanionAssignmentV2BaseSchema = z.object({
  weekday: z.number().int().min(0).max(6),
  part: z.enum(STAFF_COMPANION_PARTS),
  companion_staff_id: z.string().uuid(),
});

export type StaffCompanionAssignmentV2Base = z.infer<typeof staffCompanionAssignmentV2BaseSchema>;

export const staffCompanionAssignmentV2ReadSchema = staffCompanionAssignmentV2BaseSchema.extend({
  id: z.string().uuid(),
  trainee_staff_id: z.string().uuid(),
  created_at: z.string(),
  updated_at: z.string(),
});

export type StaffCompanionAssignmentV2Read = z.infer<typeof staffCompanionAssignmentV2ReadSchema>;

// ---------------------------------------------------------------------------
// Bulk PUT schema (same validation as backend)
//   - 0..14 items (7 曜日 × 3 part)
//   - 同 (weekday, part) 重複禁止
//   - 同曜日内: full と am/pm の併存禁止
// ---------------------------------------------------------------------------

export const staffCompanionAssignmentsBulkPutSchema = z.object({
  items: z
    .array(staffCompanionAssignmentV2BaseSchema)
    .max(14, '最大 14 件までです')
    .superRefine((items, ctx) => {
      const wkSeen: Record<number, Set<string>> = {};
      items.forEach((it, i) => {
        if (!wkSeen[it.weekday]) wkSeen[it.weekday] = new Set<string>();
        const set = wkSeen[it.weekday]!;

        // 同 (weekday, part) 重複
        if (set.has(it.part)) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            path: [i, 'part'],
            message: '同曜日内 part 重複',
          });
        }

        // full と am/pm の併存禁止
        if (
          (it.part === 'full' && (set.has('am') || set.has('pm'))) ||
          ((it.part === 'am' || it.part === 'pm') && set.has('full'))
        ) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            path: [i, 'part'],
            message: 'full と am/pm の併存不可',
          });
        }

        set.add(it.part);
      });
    }),
});

export type StaffCompanionAssignmentsBulkPut = z.infer<
  typeof staffCompanionAssignmentsBulkPutSchema
>;
